Signal/src/util/ThreadUtil.m

```diff
@@ -84,6 +88,112 @@ + (void)sendMessageWithAttachment:(SignalAttachment *)attachment
         }];
 }
 
++ (void)createBlockOfferIfNecessary:(TSContactThread *)contactThread
+                     storageManager:(TSStorageManager *)storageManager
+                    contactsManager:(OWSContactsManager *)contactsManager
+                    blockingManager:(OWSBlockingManager *)blockingManager
+{
+    OWSAssert(contactThread);
+    OWSAssert(storageManager);
+    OWSAssert(contactsManager);
+    OWSAssert(blockingManager);
+
+    if ([[blockingManager blockedPhoneNumbers] containsObject:contactThread.contactIdentifier]) {
+        // Only create block offers for users which are not already blocked.
+        return;
+    }
+
+    Contact *contact = [contactsManager contactForPhoneIdentifier:contactThread.contactIdentifier];
+    if (contact) {
+        // Only create block offers for non-contacts.
+        return;
+    }
+
+    [storageManager.dbConnection readWriteWithBlock:^(YapDatabaseReadWriteTransaction *transaction) {
+        const int kMaxOutgoingMessageCount = 10;
+
+        __block TSIncomingMessage *firstIncomingMessage = nil;
+        __block TSOutgoingMessage *firstOutgoingMessage = nil;
+        __block long outgoingMessageCount = 0;
+        __block BOOL hasUnknownContactBlockOffer = NO;
+
+        NSMutableArray *blockOffers = [NSMutableArray new];
+        [[transaction ext:TSMessageDatabaseViewExtensionName]
+            enumerateRowsInGroup:contactThread.uniqueId
+                      usingBlock:^(
+                          NSString *collection, NSString *key, id object, id metadata, NSUInteger index, BOOL *stop) {
+
+                          if ([object isKindOfClass:[OWSUnknownContactBlockOfferMessage class]]) {
+                              hasUnknownContactBlockOffer = YES;
+                              [blockOffers addObject:object];
+                          } else if ([object isKindOfClass:[TSIncomingMessage class]]) {
+                              TSIncomingMessage *incomingMessage = (TSIncomingMessage *)object;
+                              if (!firstIncomingMessage ||
```

Comment:
Our DB already ordered the objects in this extension, so we shouldn't have to do any sorting in the application, ditto for the outgoing messages.